File: packages/fangyuan-rule-java/src/main/java/com/olymtech/SortRule.java
package com.olymtech;

import java.util.Map;
import java.util.Optional;

public class SortRule<T> implements Rule<T> {
    private Map<String, Rule<T>> rules;
    private Optional<Rule<T>> rule = Optional.empty();

    SortRule(Map<String, Rule<T>> rules){
        this.rules = rules;
    }

    @Override
    public boolean evaluate(final Map<String, Object> ctx) {
        this.rule = this
                .rules
                .entrySet()
                .stream()
                .filter(rule -> rule.getValue().evaluate(ctx))
                .findFirst()
                .map(entry -> entry.getValue());
        return this.rule.isPresent();
    }

    @Override
    public T execute(){
        return this.rule.map(value -> value.execute()).orElse(null);
    }
}

File: packages/fangyuan-rule-sdk/src/SortRule.ts
import { Rule } from "./Rule";

export class SortRule<T, R = Record<string, any>> implements Rule<T> {
  rule: Rule<T> | null = null;
  constructor(
    private readonly ctx: R,
    private readonly rules: ((ctx: R) => Rule<T>)[],
    private readonly attributes: Record<string, string> = {}
  ) {}

  evaluate(): boolean {
    for (let ruleCtor of this.rules) {
      const rule = ruleCtor(this.ctx);
      if (rule.evaluate()) {
        this.rule = rule;
        return true;
      }
    }

    return false;
  }

  execute(): T {
    if (!this.rule) {
      throw new Error("请确保规则执行成功再调用");
    }
    return this.rule.execute();
  }

  toString() {
    this.attributes.name;
  }
}
